File: packages/template-python/src/deepagents_template/runtime/config/config_paths.py
"""Config path resolution — port of TS ``config-paths.ts``."""

from __future__ import annotations

import os
from pathlib import Path

from deepagents_template.runtime.config.config_schema import (
    BUILTIN_TEMPLATE_CONFIGS,
    DEFAULT_BUILTIN_TEMPLATE_CONFIG,
)
from deepagents_template.runtime.logger import logger

TEMPLATE_PACKAGE_ROOT = Path(__file__).resolve().parent.parent.parent.parent.parent


def readBuiltinTemplateConfigNameFromEnv() -> str | None:
    name = os.environ.get("DEEPAGENTS_BUILTIN_CONFIG")
    if not name:
        return None
    if name in BUILTIN_TEMPLATE_CONFIGS:
        return name
    logger.warning(
        "Unknown DEEPAGENTS_BUILTIN_CONFIG; falling back to default",
        extra={"requested": name, "available": list(BUILTIN_TEMPLATE_CONFIGS.keys())},
    )
    return None


def resolveBuiltinTemplateConfig(name: str | None = None) -> dict[str, str]:
    name = name or DEFAULT_BUILTIN_TEMPLATE_CONFIG
    cfg = BUILTIN_TEMPLATE_CONFIGS[name]
    return {
        "path": str(TEMPLATE_PACKAGE_ROOT / cfg["path"]),
        "resourceBase": str(TEMPLATE_PACKAGE_ROOT / cfg["resourceBase"]),
    }


def deepAgentsHome() -> str:
    return os.environ.get("DEEPAGENTS_HOME") or str(Path.home() / ".deepagents")


def resolveConfigResourcePath(path: str, baseDir: str) -> str:
    if path.startswith("~/") or path.startswith("~/.deepagents/") or Path(path).is_absolute():
        return path
    return str(Path(baseDir) / path)


def resolvePath(filePath: str, baseDir: str | None = None) -> str:
    base = baseDir or os.getcwd()
    if filePath == "~/.deepagents":
        return deepAgentsHome()
    if filePath.startswith("~/.deepagents/"):
        return str(Path(deepAgentsHome()) / filePath[len("~/.deepagents/"):])
    if filePath.startswith("~/"):
        return str(Path.home() / filePath[2:])
    if Path(filePath).is_absolute():
        return filePath
    return str(Path(base) / filePath)

File: packages/deepagents-flow-ts/src/runtime/config/config-paths.ts
/**
 * Config path resolution.
 *
 * Filesystem locations for the template package, the user's ~/.flowagents home,
 * and the built-in template config presets. Pure path logic — no config parsing
 * or merging. Extracted from config-loader.ts.
 */
import { homedir } from "node:os";
import { isAbsolute, join, resolve } from "node:path";
import { logger } from "../logger.js";
import { resolvePackageRoot } from "../package-root.js";
import { FLOWAGENTS_DIRNAME } from "../paths.js";
import {
  BUILTIN_TEMPLATE_CONFIGS,
  DEFAULT_BUILTIN_TEMPLATE_CONFIG,
  type BuiltinTemplateConfigName,
} from "./config-schema.js";

/** Agent 模板包根（含 config/、prompts/）；与 ACP 用户 workspace cwd 无关。 */
export const TEMPLATE_PACKAGE_ROOT = resolvePackageRoot(import.meta.url);

export function readBuiltinTemplateConfigNameFromEnv(): BuiltinTemplateConfigName | undefined {
  const name = process.env.DEEPAGENTS_BUILTIN_CONFIG;
  if (!name) {
    return undefined;
  }
  if (name in BUILTIN_TEMPLATE_CONFIGS) {
    return name as BuiltinTemplateConfigName;
  }
  logger.warn("Unknown DEEPAGENTS_BUILTIN_CONFIG; falling back to default", {
    requested: name,
    available: Object.keys(BUILTIN_TEMPLATE_CONFIGS),
  });
  return undefined;
}

export function resolveBuiltinTemplateConfig(
  name: BuiltinTemplateConfigName = DEFAULT_BUILTIN_TEMPLATE_CONFIG
): { path: string; resourceBase: string } {
  const config = BUILTIN_TEMPLATE_CONFIGS[name];
  return {
    path: resolve(TEMPLATE_PACKAGE_ROOT, config.path),
    resourceBase: resolve(TEMPLATE_PACKAGE_ROOT, config.resourceBase),
  };
}

/** ~/.flowagents —— 全局配置目录（config.json / models.json / mcp.json / skills / plugins / sessions）。 */
export function flowAgentsHome(): string {
  return resolve(process.env.FLOWAGENTS_HOME || join(homedir(), FLOWAGENTS_DIRNAME));
}


export function resolveConfigResourcePath(path: string, baseDir: string): string {
  if (path.startsWith("~/") || isAbsolute(path)) {
    return path;
  }
  return resolve(baseDir, path);
}

export function resolvePath(filePath: string, baseDir = process.cwd()): string {
  // ~/.flowagents — 全局配置根
  if (filePath === `~/${FLOWAGENTS_DIRNAME}`) {
    return flowAgentsHome();
  }
  if (filePath.startsWith(`~/${FLOWAGENTS_DIRNAME}/`)) {
    return resolve(flowAgentsHome(), filePath.slice(`~/${FLOWAGENTS_DIRNAME}/`.length));
  }
  if (filePath.startsWith("~/")) {
    return resolve(homedir(), filePath.slice(2));
  }
  if (isAbsolute(filePath)) {
    return filePath;
  }
  return resolve(baseDir, filePath);
}
